### PR TITLE
Artwr/s3 to hive operator

### PR DESCRIPTION
**Pull request by krishnabhupatiraju**
@mistercrunch @artwr Please take a look. The change is to add an operator to download a file from s3 and upload to a Hive table.

### CODE REVIEW

File: airflow/hooks/S3_hook.py [hunk 1]
```diff
@@ -105,6 +105,15 @@ def __setstate__(self, d):
         self.__dict__.update(d)
         self.__dict__['connection'] = self.get_conn()
 
+    def _parse_s3_url(self, s3url):
+        parsed_url = urlparse(s3url)
+        if parsed_url.netloc == '':
```
Comment (mistercrunch) [inline airflow/hooks/S3_hook.py]: `if not parsed_url.netloc`also catches None, pythoneskier

File: airflow/hooks/S3_hook.py [hunk 2]
```diff
@@ -187,15 +196,24 @@ def check_for_key(self, key, bucket_name=None):
         Checks that a key exists in a bucket
         '''
         if bucket_name is None:
```
Comment (mistercrunch) [inline airflow/hooks/S3_hook.py]: `if not bucket_name:`

File: airflow/hooks/S3_hook.py [hunk 3]
```diff
@@ -187,15 +196,24 @@ def check_for_key(self, key, bucket_name=None):
...
+    def get_key(self, key, bucket_name=None):
+        '''
+        Returns a boto.s3.key.Key object
+
+        :param key: the path to the key
+        :type key: str
+        :param bucket_name: the name of the bucket
+        :type bucket_name: str
+        '''
+        if bucket_name is None:
```
Comment (mistercrunch) [inline airflow/hooks/S3_hook.py]: `if not bucket_name:`

File: airflow/operators/s3_to_hive_operator.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    }
+    template_fields = ('s3_key', 'partition', 'hive_table')
+    template_ext = ()
+    ui_color = '#a0e08c'
+
+    @apply_defaults
+    def __init__(
+            self,
+            s3_key,
+            field_dict,
```
Comment (mistercrunch) [inline airflow/operators/s3_to_hive_operator.py]: field_dict is a required argument but isn't documented, you can probably just copy to 2 lines from the hooks' docstring

File: airflow/operators/s3_to_hive_operator.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+    Hive data types are inferred from the cursors's metadata from.
+
+    Note that the table genearted in Hive uses ``STORED AS textfile``
+    which isn't the most efficient serialization format. If a
+    large amount of data is loaded and/or if the tables gets
+    queried considerably, you may want to use this operator only to
+    stage the data into a temporary table before loading it into its
+    final destination using a ``HiveOperator``.
+
+    :param key: The key to be retrieved from S3
```
Comment (mistercrunch) [inline airflow/operators/s3_to_hive_operator.py]: `s3_key`

File: airflow/operators/s3_to_hive_operator.py [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+    :type key: str
```
Comment (mistercrunch) [inline airflow/operators/s3_to_hive_operator.py]: `s3_key`

File: airflow/operators/s3_to_hive_operator.py [hunk 7]
```diff
@@ -0,0 +1,98 @@
+import csv
```
Comment (mistercrunch) [inline airflow/operators/s3_to_hive_operator.py]: unused

File: airflow/operators/s3_to_hive_operator.py [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+import logging
+from tempfile import NamedTemporaryFile
+import boto
```
Comment (mistercrunch) [inline airflow/operators/s3_to_hive_operator.py]: unused

File: airflow/operators/s3_to_hive_operator.py [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+        self.hive = HiveCliHook(hive_cli_conn_id=hive_cli_conn_id)
+        self.s3 = S3Hook(s3_conn_id=s3_conn_id)
+
+    def execute(self, context):
+        logging.info("Downloading S3 file")
+        if not self.s3.check_for_key(self.s3_key):
+            raise Exception("The key {0} does not exists".format(self.s3_key))
+        s3_key_object = self.s3.get_key(self.s3_key)
+        with NamedTemporaryFile("w") as f:
+            logging.info("Dumping S3 file {0} contents to local file {1}".format(self.s3_key, f.name))
```
Comment (mistercrunch) [inline airflow/operators/s3_to_hive_operator.py]: break down to 80- width

File: airflow/hooks/S3_hook.py [hunk 10]
```diff
@@ -206,3 +224,18 @@ def check_for_prefix(self, bucket_name, prefix, delimiter):
         leaf_prefix = prefix_split[1]
         plist = self.list_prefixes(bucket_name, previous_level, delimiter)
         return False if plist is None else prefix in plist
+
+    def load_file(self, filename,
+                  key, bucket_name=None,
+                  replace=False):
+        if bucket_name is None:
```
Comment (mistercrunch) [inline airflow/hooks/S3_hook.py]: `if not bucket_name:`

File: airflow/hooks/S3_hook.py [hunk 11]
```diff
@@ -206,3 +224,18 @@ def check_for_prefix(self, bucket_name, prefix, delimiter):
         leaf_prefix = prefix_split[1]
         plist = self.list_prefixes(bucket_name, previous_level, delimiter)
         return False if plist is None else prefix in plist
+
+    def load_file(self, filename,
```
Comment (mistercrunch) [inline airflow/hooks/S3_hook.py]: + docstring & doctest please

**Comment by mistercrunch** [comment]
Thanks for writing this! 

We should add a unitest as part of this PR too.

**Comment by artwr** [comment]
We will need to talk about unit testing because of the difficulty of testing with S3 + credentials. Maybe something like fakes3?

**Comment by mistercrunch** [comment]
Can we get PI to create a test bucket? There might be one already.

**Comment by artwr** [comment]
Closing this one but leaving it for posterity.

**Pull request closed without merging**